### PR TITLE
Ng 49 cleanup optimization

### PR DESCRIPTION
**Pull request by cfboy**

### CODE REVIEW

File: GUI/semref.kv [hunk 1]
```diff
@@ -28,64 +28,64 @@
     canvas.before:
         
         Color:
-            rgb: 1,.8,0
+            rgb: self.border
```
Comment (ChristianPerez34) [inline GUI/semref.kv]: `rgb: self.border_color`

File: GUI/semref.kv [hunk 2]
```diff
@@ -95,128 +95,133 @@
...
             size: dp(40), dp(40) 
         Rectangle: 
-            pos: dp(620),dp(20)
+            pos: self.box_pos_x + dp(320), self.box_pos_y + dp(5)
             size: dp(40), dp(40)  
             
 <SevenSegmentDisplay>
     canvas.before:
+        Color:
+            rgb: self.border
```
Comment (ChristianPerez34) [inline GUI/semref.kv]: `rgb: self.border_color`

File: GUI/window.py [hunk 3]
```diff
@@ -45,6 +45,7 @@
 editor_saved = False
 cleared = True
 is_obj = False
+theme_manager = ThemeManager()
```
Comment (ChristianPerez34) [inline GUI/window.py]: Can be accessed from from application. Should not be a global variable

File: GUI/window.py [hunk 4]
```diff
@@ -969,13 +969,18 @@ def get_data(self, address, instruction):
 
 
 class TrafficLights(Widget):
+    global theme_manager
```
Comment (ChristianPerez34) [inline GUI/window.py]: Can be accessed from app. Not needed

File: GUI/window.py [hunk 5]
```diff
@@ -969,13 +969,18 @@ def get_data(self, address, instruction):
...
     red_1 = ListProperty([1, 0, 0])
     red_2 = ListProperty([1, 0, 0])
     yellow_1 = ListProperty([1, 1, 0])
     yellow_2 = ListProperty([1, 1, 0])
     green_1 = ListProperty([0, 1, 0])
     green_2 = ListProperty([0, 1, 0])
 
+    box_pos_x = dp(850)
+    box_pos_y = dp(120)
+    border = ListProperty(theme_manager.primary_color)
```
Comment (ChristianPerez34) [inline GUI/window.py]: should be inside `__init__`

File: GUI/window.py [hunk 6]
```diff
@@ -1064,6 +1069,7 @@ def change_color(self, binary):
 
 
 class SevenSegmentDisplay(Widget):
+    global theme_manager
```
Comment (ChristianPerez34) [inline GUI/window.py]: Can be accessed from app. Not needed

File: GUI/window.py [hunk 7]
```diff
@@ -1079,6 +1085,12 @@ class SevenSegmentDisplay(Widget):
     rightE = ListProperty([.41, .41, .41])
     rightF = ListProperty([.41, .41, .41])
     rightG = ListProperty([.41, .41, .41])
+    border = ListProperty(theme_manager.primary_color)
```
Comment (ChristianPerez34) [inline GUI/window.py]: theme manager can be accessed from app. You might need to move this inside of the `__init__` method

`border_color = ListProperty(theme_manager.primary_color)`

File: utils.py [hunk 8]
```diff
@@ -12,6 +12,12 @@ def convert_to_hex(num, bits):
 
 
 def convert_to_binary(num, bits):
+    """
+    Converts int number to hexadecimal
```
Comment (ChristianPerez34) [inline utils.py]: Converts number to its binary representation

File: utils.py [hunk 9]
```diff
@@ -12,6 +12,12 @@ def convert_to_hex(num, bits):
...
+    :param num:
+    :param bits:
+    :return: str:
```
Comment (ChristianPerez34) [inline utils.py]: :param num: int
      :param bits: str
      :return: str

File: utils.py [hunk 10]
```diff
@@ -161,34 +173,35 @@ def update_indicators(instance, is_file_loaded):
...
     'port': 3
 }
-
+# General Information of HEX Keyboard Object.
 HEX_KEYBOARD = {
     'menu_title': 'Configure Hex Keyboard Port',
     'port': 2
 }
 
+# This ports are reserved by a I/O devices.
```
Comment (ChristianPerez34) [inline utils.py]: Reserved ports for I/O devices.

**Review by ChristianPerez34 — CHANGES_REQUESTED** [review]
Pull from master. Will comment further recommendations once branch is up-to-date

File: GUI/window.py [hunk 11]
```diff
@@ -12,7 +15,7 @@
 from kivy.graphics.context_instructions import Color
 from kivy.graphics.vertex_instructions import Line, Rectangle
 from kivy.metrics import MetricsBase, dp, sp
-from kivy.properties import ListProperty
+from kivy.properties import ListProperty, get_color_from_hex
```
Comment (ChristianPerez34) [inline GUI/window.py]: from kivy.utils import get_color_from_hex

Must also run the pycharm optimize imports feature

File: utils.py [hunk 12]
```diff
@@ -65,17 +70,22 @@ def update_reserved_ports(device, port_to_remove, port_to_add, reserve_block=Fal
 
 
 def update_indicators(instance, is_file_loaded):
-
+    """
+    This method manage Loaded/NotLoaded File indicators using a condition.
+    :param instance:
+    :param is_ram_loaded:
```
Comment (ChristianPerez34) [inline utils.py]: :param instance: obj
:param is_ram_loaded: bool

**Review by ChristianPerez34 — CHANGES_REQUESTED** [review]
Doesn't conform to upper case naming style
```Python
file_path = ''
can_write = False
loaded_file = False
run_editor = True
editor_saved = False
cleared = True
is_obj = False
```
`HexKeyboard`
simplify if statements

```Python
            for i in range(16):
                if i < 4:
                    Line(rectangle=(dp(340 + (89 * (i % 4))),
                                    dp(357), dp(87), dp(35)), width=dp(0.8))
                elif i >= 4 and i < 8:
                    Line(rectangle=(dp(340 + (89 * (i % 4))),
                                    dp(322), dp(87), dp(35)), width=dp(0.8))
                elif i >= 8 and i < 12:
                    Line(rectangle=(dp(340 + (89 * (i % 4))),
                                    dp(287), dp(87), dp(35)), width=dp(0.8))
                else:
                    Line(rectangle=(dp(340 + (89 * (i % 4))),
                                    dp(252), dp(87), dp(35)), width=dp(0.8))
```

`def write_ram(self)`, `def check_loader(self, dt)`, `def open_keyboard(self, instance)`, `def update_io(self, dt)`, `def run_micro_instructions(self, instance)`, `def run_micro_instructions_step(self, instance)`, `def assembler(self)`, `def run_micro_sim(self, file)` missing docstring

`def run_micro_instructions` edit for loop to a while loop since m is not being used
```Python
                for m in range(2):
                    if self.first_inst:
                        self.run_window.inst_table.data_list.clear()
                        self.run_window.inst_table.get_data(self.micro_sim.index,
                                                            self.micro_sim.disassembled_instruction())
                        self.run_window.inst_table.get_data(self.micro_sim.index,
                                                            self.micro_sim.disassembled_instruction())
                        self.first_inst = False
                    else:
                        self.micro_sim.prev_index = -1
                        self.run_window.blinking_on.cancel()
                        self.run_window.blinking_off.cancel()

                        self.run_window.blinking_on()
                        self.run_window.blinking_off()

                        timeout = time.time() + 5  # 5 seconds from now
                        while self.micro_sim.is_running:
                            try:
                                self.micro_sim.run_micro_instructions(timeout)
                                self.run_window.inst_table.get_data(self.micro_sim.index,
                                                                    self.micro_sim.disassembled_instruction())

                                if self.micro_sim.prev_index == self.micro_sim.index:
                                    self.micro_sim.is_running = False
                                else:
                                    self.micro_sim.prev_index = self.micro_sim.index
                            except (SystemError, TimeoutError) as e:
                                self.micro_sim.is_running = False
                                toast_message = f'Error! {e}'
                    self.run_window.reg_table.get_data()
                    self.run_window.mem_table.data_list.clear()
                    self.run_window.mem_table.get_data()
                    toast(toast_message)
```

`def assembler(self)` m is not being used. Change for loop to a while loop
```Python
            for m in range(50):
                f.write(f'{RAM_ASSEMBLER[i]} {RAM_ASSEMBLER[i + 1]}' + '\n')
                i += 2
```

**Comment by ChristianPerez34** [comment]
`def clear_dialog(self, instance)`, `def clear_decision(self, *args)`, `def clear(self)`, `def clear_run(self)`, `def open_editor_save_dialog(self, instance)`, `def save_asm_file(self, *args)`, `def buttons_information(self, instance)`, `def io_config_open(self, instance)`, `def save_io_ports(self, *args)`, `def file_manager_open(self, instance)`, `def get_data(self)` missing docstring

`def save_file` m is not used. Change for loop to a while loop
```Python
            for m in range(50):
                f.write(f'\n{RAM[i]}    {RAM[i + 1]}')
                i += 2
```

**Review by noelvalentin — APPROVED** [review]

**Review by ChristianPerez34 — APPROVED** [review]

**Pull request merged**